### DISCUSSION TITLE
Pi & ESP Wiring Question

### DISCUSSION THREAD

**Discussion by realroywalker** [awawa-dev/HyperHDR#399] [Q&A] [answered]
Hi all, 
I currently run Hyperion on a Pi Zero and use WLED to drive my SK6812 strip on an ESP8266, but I'd like to revamp my setup after finding HyperHDR.
I would like to switch to wired comms between the Pi and ESP8266 as part of this, from what I can understand reading through the github pages, the way to do this is flash HyperSPI on the ESP and then connect the MOSI/MISO and SCLK between the Pi/ESP.

I would however like to keep my Pi power supply separate (for stability mainly, and so that I'm not using valuable juice from my 10a supply that I have to drive my LEDs)

My question, is will this cause issues with grounds etc, given that the ESP will be connected to the Pi via the MOSI/MISO and SCLK connections?

This is a crude drawing of how I imagine it in my head, but i'm not sure if it would work?

![hyperhdr](https://user-images.githubusercontent.com/486860/199767315-b7528331-c54b-48b9-9a29-670d8e1f7702.png)

Any advice on how I can hook these up is appreciated, or some way to isolate the devices (if needed) so that I can use separate power for each.

**Comment by awawa-dev** [comment]
Hi
1. If you provide power supply via pin-out then you also bypass fuse and power protection circuit for RPI and probably diode for ESP board. I wouldn't do that and instead try a bit simpler approach. Use one solid 2A USB power supply (and cable), connect it to the Rpi. Then connect ESP using USB to the Rpi for power supplying. This should provide a common ground between Rpi & ESP, you need only one additional cable to provide a common ground between ESP and the LED strip
2. If you have an USB grabber that you want to connect it to Rpi and you don't have USB hub then probably you may have one free Rpi USB port missing. Still an USB hub would be a better solution but you can power ESP (+5V and the ground) from the LED strip but to ESP's USB port instead of the pinout. Or you can try pinout power supply for ESP avoiding the diode: it's a bit risky but ESP costs much less than Rpi and Rpi rather should survive if something goes wrong.

**Comment by awawa-dev** [comment, marked as answer]
and similar to point 1)  

3. use Rpi: +5V and ground from Rpi's pinout and connect it to the ESP USB as a power supply for ESP. This should be stabilize by Rpi so you may also connect it to the ESP pinout as an alternative power supply with little risk.

**Reply by realroywalker** [reply]
@awawa-dev thanks so much for the tips, I've drawn it out (using fritzing this time, what a great app!) hopefully I have understood your suggestions correctly?

My only concern is around the LED strip - can I simply feed power from the 10a supply like that, or does GND of the LED strip need to be tied to the ESP somehow?
 
![hyperhdr_bb](https://user-images.githubusercontent.com/486860/199961404-7180f854-3a05-4f83-9f32-eb0ce9ebc293.png)

**Reply by awawa-dev** [reply]
Nice scheme :) 👍 
You are right. You need to connect LED strip ground with the ESP pinout ground. The only question is if it work stable without a level shifter... for most user it works but you must remember, there is a possibility that you may need it later: https://github.com/awawa-dev/HyperHDR/wiki/Level-Shifter because ESP (and Rpi also) has a 3.3V pinout for the data line and the LED strip expects 5V.

**Reply by realroywalker** [reply]
Thanks for the help, simply connecting the ESP ground like ? - it causes no issues having the 10a supply ground connected too?

![hyperhdr_bb](https://user-images.githubusercontent.com/486860/199965729-f4c2bdf5-ec40-41e7-9ce2-ed34c20bdbb3.png)

I will checkout the logic level shifter, I also read that the sacrificial pixel method could work too so might give that a go.

**Reply by awawa-dev** [reply]
You don't need to solder the ESP ground to the same point on the LED strip where you connect the power supply's ground: you may connect it to the power supply ground somewhere near the LED strip or you can use additional ground output if your power supply has them.

Like on this scheme (LED strip ground is marked as '-V' here).
![obraz](https://user-images.githubusercontent.com/69086569/199968904-54963178-839a-4492-b13d-2d0298a99ee7.png)

**Reply by realroywalker** [reply]
Thank you, just been testing everything out on a breadboard setup and it works great. 
I do have one question though - I was looking at using a sacrificial LED as the logic shifter, so I compiled HyperSPI firmware with the skip first LED option enabled, this works (the first led isn't lit when outputting a rainbow effect as a test). However, I have setup a small test strip of 10 leds (with number 1 being the LED to sacrifice) so I configure HyperHDR to output 9 LEDs in my arrangement like so ;

![image](https://user-images.githubusercontent.com/486860/199999132-1138d31d-939e-4c21-9bc7-97cf0761158a.png)

In this setup, only 8 of the 10 LEDs are lit, (first one skipped, then the following 8 light up).. if i change the configuration in that above screenshot to 10 then 9 leds are lit, but this obviously causes problems with lining placement up, since i've only actually got 9 active LEDs.. am I missing something obvious about this option?

**Reply by awawa-dev** [reply]
That's right. Even if you set skip the first LED in the driver firmware, neopixel sk6812/ws2812b LED strip protocol still requires to include a data for the first/skipped LED: we can't omit it. The driver firmware simply guards that the first LED won't lit up which could cause a problem if the sacrificial LED has a limited power supply from the Rpi or ESP directly (usually only few mA are allowed for Rpi).  

So need to include that LED in the HyperHDR LED strip geometry design even if you don't use it. It shouldn't be problematic if the LED strip begins somewhere in the middle of TV edge.

**Reply by awawa-dev** [reply]
BTW with this PR you don't need special firmware: https://github.com/awawa-dev/HyperHDR/pull/379
You can disable a LED and HyperHDR will send a black color for disabled LEDs.

**Reply by realroywalker** [reply]
Thanks for the explanation, looking at the PR if the LED is disabled its still included in the geometry? - maybe it would be good if there was a tick box in the HyperHDR LED instance to mark if a sacrificial LED is used ? 
This way, it would know to send some color data to the first sacrificial LED, but the following LEDs could be accurately placed / displayed in the geometry, and just have their data shifted (n +1) when it's sent?

**Reply by awawa-dev** [reply]
The main purpose was to make possible to disable a whole segment of LEDs if they are not used (for example behind TV mounting). Without this new option we had to cut it off and join remaining segments. Sacrificial LED is a special case but will also benefit from this PR (and I can simplify HyperSerial firmwares also). Still using SN74AHCT125N as a level shifter is a preferable solution. I use it over a year without any problems.

**Reply by realroywalker** [reply]
Ah ok got you, I will take a look at a proper level shifter then - I was just going with the sacrificial led since I had them to hand 😄
